src/Passes/README.md

```diff
@@ -0,0 +1,322 @@
+makes sense for classical programs is to count instructions used to implement functions. Take the C program:
+
+```c
+int foo(int x)
+{
+  return x;
+}
+
+void bar(int x, int y)
+{
+  foo(x + y);
+}
+
+int main()
+{
+  foo(2);
+  bar(3, 2);
+
+  return 0;
+}
+```
+
+which produces follow IR (without optimisation):
+
+```language
+define dso_local i32 @foo(i32 %0) #0 {
+  %2 = alloca i32, align 4
+  store i32 %0, i32* %2, align 4
+  %3 = load i32, i32* %2, align 4
+  ret i32 %3
+}
+
+define dso_local void @bar(i32 %0, i32 %1) #0 {
+  %3 = alloca i32, align 4
+  %4 = alloca i32, align 4
+  store i32 %0, i32* %3, align 4
+  store i32 %1, i32* %4, align 4
+  %5 = load i32, i32* %3, align 4
+  %6 = load i32, i32* %4, align 4
+  %7 = add nsw i32 %5, %6
+  %8 = call i32 @foo(i32 %7)
+  ret void
+}
+
+define dso_local i32 @main() #0 {
+  %1 = alloca i32, align 4
+  store i32 0, i32* %1, align 4
+  %2 = call i32 @foo(i32 2)
+  call void @bar(i32 3, i32 2)
+  ret i32 0
+}
+```
+
+A stat pass for this code, would collect following statisics:
+
+```text
+Stats for 'foo'
+===========================
+Opcode          # Used
+---------------------------
+load            1
+ret             1
+alloca          1
+store           1
+---------------------------
+
+Stats for 'bar'
+===========================
+Opcode          # Used
+---------------------------
+load            2
+add             1
+ret             1
+alloca          2
+store           2
+call            1
+---------------------------
+
+Stats for 'main'
+===========================
+Opcode          # Used
+---------------------------
+ret             1
+alloca          1
+store           1
+call            2
+---------------------------
+```
+
+**Example 3: Code validation**. A third use case is code validation. For example, one could write a pass to check whether bounds are exceeded on static arrays [2].
+Note that this is a non-standard usecase as such analysis is usually made using the AST rather than at the IR level.
+
+**References**
+[1] https://github.com/banach-space/llvm-tutor#analysis-vs-transformation-pass
+[2] https://github.com/victor-fdez/llvm-array-check-pass
+
+## Out-of-source Pass
+
+This library is build as set of out-of-source-passes. All this means is that we will not be downloading the LLVM repository and modifying this repository directly. You can read more [here](https://llvm.org/docs/CMake.html#cmake-out-of-source-pass).
+
+# Getting started
+
+## Dependencies
+
+This library is written in C++ and depends on:
+
+- LLVM
+
+Additional development dependencies include:
+
+- CMake
+- clang-format
+- clang-tidy
+
+## Building the passes
+
+To build the passes, create a new build directory and switch to that directory:
+
+```sh
+mkdir Debug
+cd Debug/
+```
+
+To build the library, first configure CMake from the build directory
+
+```sh
+cmake ..
+```
+
+and then make your target
+
+```sh
+make [target]
+```
```

Comment:
What are possible targets and which is the default?

Reply:
There is no default target. I have added a line explaining what valid targets are.

Reply:
The fragment `make [target]` tells me that specifying the `target` is optional (we can just run `make`). 
What will happen if we run just `make`? (that's what I mean with "default")
E.g. "If `target` is not specified then ...".

Reply:
It was incorrect what I wrote. The default target is `all` - that is if you just run make, it will make all. I've updated this in #1098 with following line after the code block: 

> The default target is `all`. Other valid targets are the name of the folders in `libs/` found in the passes root.